src/controlplane/scraper.go

```diff
@@ -0,0 +1,230 @@
+		if err := opt(s); err != nil {
+			return nil, fmt.Errorf("applying config option #%d: %w", i, err)
+		}
+	}
+
+	// TODO If this could change without a restart of the pod we should run it each time we scrape data,
+	// possibly with a reasonable cache Es: NewCachedDiscoveryClientForConfig
+	s.k8sVersion, err = providers.K8s.Discovery().ServerVersion()
+	if err != nil {
+		return nil, fmt.Errorf("fetching K8s version: %w", err)
+	}
+
+	s.buildDiscoverer(s.buildComponents())
+
+	return s, nil
+}
+
+// Run scraper collect the data populating the integration entities
+func (s *Scraper) Run(i *integration.Integration) error {
+	for _, component := range s.components {
+
+		pod, err := s.discoverPod(component)
+		if err != nil {
+			return fmt.Errorf("control plane component %s discovery failed: %v", component.Name, err)
+		}
+
+		if pod == nil {
+			s.logger.Debugf("No pod found for component: %s", component.Name)
+			continue
+		}
+
+		grouper := grouper.New(
+			s.deprecatedClient(component, pod.Name),
+			component.Queries,
+			s.logger,
+			pod.Name,
+		)
+
+		job := scrape.NewScrapeJob(string(component.Name), grouper, component.Specs)
+
+		s.logger.Debugf("Running job: %s", job.Name)
+
+		result := job.Populate(i, s.config.ClusterName, s.logger, s.k8sVersion)
+
+		if len(result.Errors) > 0 {
+			s.logger.Infof("Error populating data from %s: %v", job.Name, result.Error())
+		}
+	}
+
+	return nil
+}
+
+func (s *Scraper) deprecatedClient(c Component, podName string) client.HTTPClient {
+	timeout := 500 * time.Millisecond
+
+	authMethod := controlplaneClient.None
+	// Let mTLS take precedence over service account
+	switch {
+	case c.UseMTLSAuthentication:
+		authMethod = controlplaneClient.MTLS
+	case c.UseServiceAccountAuthentication:
+		authMethod = controlplaneClient.ServiceAccount
+	default:
+		authMethod = controlplaneClient.None
+	}
+
+	return controlplaneClient.New(
+		authMethod,
+		c.TLSSecretName,
+		c.TLSSecretNamespace,
+		s.logger,
+		s.K8s,
+		c.Endpoint,
+		c.SecureEndpoint,
+		s.config.NodeIP,
+		podName,
+		c.InsecureFallback,
+		timeout,
+	)
+}
+
+func (s *Scraper) discoverPod(c Component) (*corev1.Pod, error) {
+	var discoveredPod *corev1.Pod
+	// looks for the pod match a set of defined labels
+	for _, l := range c.Labels {
+		podLister, ok := s.PodListerByNamespace[c.Namespace]
+		if !ok {
+			return nil, fmt.Errorf("pod lister for namespace: %s not found for component: %s", c.Namespace, c.Name)
+		}
+
+		selector := labels.SelectorFromSet(labels.Set(l))
+		pods, err := podLister.List(selector)
+		if err != nil {
+			return nil, fmt.Errorf("fail to list pods for component:%v selector: %v", c.Name, l)
+		}
+
+		// Validation of returned pods.
+		for _, pod := range pods {
+			if pod.Spec.NodeName != s.config.NodeName {
+				s.logger.Debugf("discarding pod: %s running outside the node", pod.Name)
+				continue
+			}
+			discoveredPod = pod
+			break
+		}
+
+	}
+
+	if discoveredPod == nil {
+		return nil, nil
```

Comment:
should be this an error? Since the user can opt in/out each component we might decide to fail if we are not able to retrieve data

Reply:
this is the main issue of this func, the result could be pod, error , or no pod found but no error. Mainly because if there is an error on the lister i think we need to stop with error other wise issues related to configuration will just be noted on debug logs.